Add Emoji 14 Support

kraftbj opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->
Adds Twemoji 14.0.1 support.
Update the regex (output of `grunt precommit:emoji`)
Updates the CDN URL to 14.0.1 (still needs uploading on the w.org side)
Updates the unit tests.
NEEDED: Detection script update.

Emoji 14 does not add any ZWJ-based emoji, unlike the past standards, so our script to detect needs a bit more work than usual. I didn't think about this and most of my computers are already upgraded to supported OS versions, so will get an old computer out of the closet to help figure out a good way to 
detect it with as little change to the code as feasible.

Trac ticket: https://core.trac.wordpress.org/ticket/55395

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

kraftbj committed:
Bump Twemoji 14.0.1

kraftbj committed:
Update Emoji CDN to new 14.0.1 directory (needed still on w.org)

kraftbj committed:
Output of precommit:emoji

kraftbj committed:
Placeholder for emoji detection

costdev commented on src/wp-includes/formatting.php:
```diff
@@ -5771,7 +5771,7 @@ function _print_emoji_detection_script() {
 		 * @param string $url The emoji base URL for svg images.
 		 */
-		'svgUrl'  => apply_filters( 'emoji_svg_url', 'https://s.w.org/images/core/emoji/13.1.0/svg/' ),
+		'svgUrl'  => apply_filters( 'emoji_svg_url', 'https://s.w.org/images/core/emoji/14.0.1/svg/' ),
```
In `tests/phpunit/tests/formatting/emoji.php`, line 9 and 10 should also be updated to change `13.1.0` to `14.0.1`.

kraftbj replied:
Ah, thank you. Updated in 3e9014396a

kraftbj committed:
Update unit tests

peterwilsoncc commented:
Some of the handshake emoji include ZWJ (U+200D), for example https://emojipedia.org/handshake-light-skin-tone-medium-light-skin-tone/

desrosj committed:
Merge branch 'trunk' of github.com:WordPress/wordpress-develop into add/emoji-14

# Conflicts:
#	package-lock.json

desrosj committed:
Use 14.0.0 for the CDN folder name.

14.0.1 and 14.0.2 does not contain any image altering changes. Use 14.0.0 to reflect the version of the images.

desrosj committed:
Twemoji 14.0.2 has been released.

desrosj committed:
Update with 14.0 specific ZWD tests.

desrosj commented:
I've made a few updates!

- Synced with `trunk`.
- I changed the URLs to be `14.0.0` instead of `14.0.1`. Since these new emoji were introduced in 14.0.0, I think it makes more sense to use that version for this update.
- Updated to [14.0.2](https://github.com/twitter/twemoji/releases/tag/v14.0.2).
- I opted to use the [Handshake: Light Skin Tone, Dark Skin Tone](https://emojipedia.org/handshake-light-skin-tone-dark-skin-tone/). I've updated the necessary locations.

I've tested this on my machine (which does not yet have Emoji 14 support) by replacing the Core CDN URLs with the GitHub ones for the Twemoji repo (https://github.com/twitter/twemoji/raw/v14.0.2/assets/), and all seems to be working as expected!

kraftbj commented:
Thanks for the updated patch! I've confirmed it does replace on my older device that doesn't support Emoji 14 and does not replace on my upgraded device with 14 support.

We need to coordinate with a CDN committer to get the assets added, but once that's done, this is good to go.

peterwilsoncc commented on src/js/_enqueues/lib/emoji-loader.js:
```diff
@@ -115,23 +115,26 @@
 				 *
-				 * 0x2764, 0xfe0f == Red Heart emoji.
-				 * 0x200D == Zero-Width Joiner (ZWJ) that links the two code points for the new emoji or
+				 * 0x1FAF1 == Rightwards Hand
```
Using the rightward hand emoji may not work for these tests as it was introduced in Emoji 14.

> When updating this test for future Emoji releases, ensure that individual emoji that make up the sequence come from older emoji standards.

I guess the reason is that some systems may end up comparing `[broken emoji]` to `[broken emoji]` and returning a false positive.

As I didn't write the comment, guess is the key word though.

kraftbj replied:
I was concerned about that initially (and why I didn't put anything in there originally; there are no new emoji that are composite of only older emoji), but it did accurately catch for me.

Once we get the assets on the CDN, we can use Browserstack or something similar to test a wider range of older platforms as confirmation.

(Edit: I'm not opposed to landing this as is and performing dedicated testing during the beta phases too).

peterwilsoncc replied:
Let's land and test during the beta. 

(Edit: I've requested the CDN be updated in the meta slack channel)

kraftbj replied:
I should have ability to commit to CDN soon. I'll update the assets in the next few hours and will report here when done.

kraftbj committed:
Fix package-lock.json

kraftbj committed:
Retry package-lock.json after nvm to node 14

peterwilsoncc commented:
Merged in https://core.trac.wordpress.org/changeset/53150, thanks.

I resolved the weird package-lock issue by ignoring the file in this PR and letting NPM do the work.
